src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -305,6 +305,13 @@ bool LibraryCallKit::arch_supports_vector(int sopc, int num_elem, BasicType type
       }
     }
 
+    // Since operation support check is done upfront, we reach here only when target
+    // specific checks in match_rule_supported_vector returned a true value
+    // for these masked operations over non-predicated targets.
+    if (!is_supported && (sopc == Op_StoreVectorMasked || sopc == Op_LoadVectorMasked)) {
+      return true;
+    }
```

Comment:
Still unclear for me. As I understand `the upfront checks`, you mention, are checks at lines 270 and 286. So we know that `StoreVectorMasked` and `LoadVectorMasked` are supported.
The second part of your comment is talking about `non-predicated targets`. Does it mean that the real check should be next?:
```
   if (Matcher::has_predicated_vectors()) {
      ....
   } else if (sopc == Op_StoreVectorMasked || sopc == Op_LoadVectorMasked) {
     // your comment
     return true;
   }
   ```
   Or I am still missing what are you trying to do here?

Reply:
Could we change the `VectorMaskUseType` here https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vectorIntrinsics.cpp#L1221 instead of doing the modification here? Originally we assume `LoadVectorMasked/StoreVectorMasked` is implemented with predicated feature, we added the `VectorMaskUsePred`. If the assumption is not changed now, we can just remote `VectorMaskUsePred` and so only `match_rule_supported_vector` is checked for such ops.

Reply:
> 

Hi Vladimir,
L220 is checking for backend support for these IR using match_rule_supported_vector, if targets do not support them then we exit early,  LoadVectorMasked and StoreVectorMasked are special IR node which were added only for predicated targets.

I agree with @XiaohongGong that we can remove this constraint for masked memory operations from the caller side i.e. LibraryCallKit::inline_vector_mem_masked_operation,  such a constraint is only useful for shared IR nodes which can carry additional mask edge on common IR for predicated targets, since there we need to check for existence of blend + vector op as a fall back case.